### DISCUSSION TITLE
Splink and Python 3.11.3

### DISCUSSION THREAD

**Discussion by DOH-REP1303** [moj-analytical-services/splink#1481] [Q&A] [answered]
Hello.
I just upgraded to the latest version of Anaconda because I was having trouble running my code from the command prompt. The error message was that my python versions did not match. So I installed Anaconda3-2023.07-1-Windows-x86_64. The installation was fine. Before I upgraded Anaconda, the code ran fine and was generating predictions. The code starts executing and then generates the following message: 
Traceback (most recent call last):
  File "C:\Users\rep1303\AppData\Local\anaconda3\Lib\site-packages\splink\linker.py", line 643, in _log_and_run_sql_execution
    return self._run_sql_execution(final_sql, templated_name, physical_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rep1303\AppData\Local\anaconda3\Lib\site-packages\splink\duckdb\linker.py", line 225, in _run_sql_execution
    self._con.execute(final_sql)
duckdb.OutOfRangeException: Out of Range Error: cannot take logarithm of zero

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Python\ESR_Project\PythonMatchCode_Version_2.0.py", line 239, in <module>
    df_predictions = linker.predict(threshold_match_probability=0.2)

Normally I would conclude that there is a data issue. However, I have not changed any data since doing the upgrade. Any ideas?

I have attached my python code as a txt file. I am running python 3.11.3 from Spyder. I get the same result running the code in Spyder, from the command prompt and from the Anaconda prompt.

Thank you for your help.

[PythonMatchCode_Version_2.0.txt](https://github.com/moj-analytical-services/splink/files/12186156/PythonMatchCode_Version_2.0.txt)

**Comment by RobinL** [comment]
Could be a small sample thing, try running again or upping target rows to say 1e7 in the following:
linker.estimate_u_using_random_sampling(target_rows=1e6)

**Reply by DOH-REP1303** [reply]
Thank you for your help. Changing to 1e7 did not help. Same error as before.

**Reply by RobinL** [reply]
Hmm.  Could you save the model to json immediately before calling predict (but after training) using `linker.save_model_to_json(path)` [i.e. this](
https://github.com/moj-analytical-services/splink/blob/1acc6298eea6ee514a8355c34add21c559ce899d/splink/linker.py#L3083)
and post the results here to help me debug.

The file just contains the model, not the data, but feel free to verify it contains nothing sensitive before posting.

**Reply by RobinL** [reply, marked as answer]
You could also try making a revision like the following (see the top of the page where i use [probability_two_random_records_match](https://moj-analytical-services.github.io/splink/demos/tutorials/04_Estimating_model_parameters.html#estimation-of-probability_two_random_records_match)).  It generally makes the model more stable.  


https://gist.github.com/RobinL/5515d23f98fb13b5f6a3f423e5f15fa9/revisions

**Reply by DOH-REP1303** [reply]
Here are the results.
[my_settings.txt](https://github.com/moj-analytical-services/splink/files/12186575/my_settings.txt)

**Reply by DOH-REP1303** [reply]
I made the revisions suggested above and got the error below. I am wondering if I am missing a module.
C:\Users\rep1303>C:\Users\rep1303\AppData\Local\anaconda3\python.exe C:\Python\ESR_Project\PythonMatchCode_Version_2.0.py
Traceback (most recent call last):
  File "C:\Users\rep1303\AppData\Local\anaconda3\Lib\site-packages\splink\linker.py", line 643, in _log_and_run_sql_execution
    return self._run_sql_execution(final_sql, templated_name, physical_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rep1303\AppData\Local\anaconda3\Lib\site-packages\splink\duckdb\linker.py", line 225, in _run_sql_execution
    self._con.execute(final_sql)
duckdb.ParserException: Parser Error: syntax error at or near "and"
LINE 13: ...   CREATE TABLE __splink__df_count_cumulative_blocks_786f2c00a
        AS
    ...
                                                   ^

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Python\ESR_Project\PythonMatchCode_Version_2.0.py", line 202, in <module>
    linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.7)
  File "C:\Users\rep1303\AppData\Local\anaconda3\Lib\site-packages\splink\linker.py", line 3148, in estimate_probability_two_random_records_match
    records = cumulative_comparisons_generated_by_blocking_rules(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rep1303\AppData\Local\anaconda3\Lib\site-packages\splink\analyse_blocking.py", line 100, in cumulative_comparisons_generated_by_blocking_rules
    cumulative_blocking_rule_count = linker._execute_sql_pipeline([concat])
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rep1303\AppData\Local\anaconda3\Lib\site-packages\splink\linker.py", line 581, in _execute_sql_pipeline
    raise e
  File "C:\Users\rep1303\AppData\Local\anaconda3\Lib\site-packages\splink\linker.py", line 575, in _execute_sql_pipeline
    dataframe = self._sql_to_splink_dataframe_checking_cache(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rep1303\AppData\Local\anaconda3\Lib\site-packages\splink\linker.py", line 841, in _sql_to_splink_dataframe_checking_cache
    splink_dataframe = self._execute_sql_against_backend(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rep1303\AppData\Local\anaconda3\Lib\site-packages\splink\duckdb\linker.py", line 220, in _execute_sql_against_backend
    self._log_and_run_sql_execution(sql, templated_name, physical_name)
  File "C:\Users\rep1303\AppData\Local\anaconda3\Lib\site-packages\splink\linker.py", line 655, in _log_and_run_sql_execution
    raise SplinkException(
splink.exceptions.SplinkException: Error executing the following sql for table `__splink__df_count_cumulative_blocks`(__splink__df_count_cumulative_blocks_786f2c00a):

        CREATE TABLE __splink__df_count_cumulative_blocks_786f2c00a
        AS
        (WITH __splink__df_concat as (select * from __splink__df_concat_a2bd1b3b9),
__splink__df_blocked_data as (
            select
            "l"."source_dataset" as "source_dataset_l", "r"."source_dataset" as "source_dataset_r", "l"."unique_id" as "unique_id_l", "r"."unique_id" as "unique_id_r", "l"."first_name" as "first_name_l", "r"."first_name" as "first_name_r", "l"."last_name" as "last_name_l", "r"."last_name" as "last_name_r", "l"."mi" as "mi_l", "r"."mi" as "mi_r", "l"."dob" as "dob_l", "r"."dob" as "dob_r"
            , '0' as match_key

            from __splink__df_concat as l
            inner join __splink__df_concat as r
            on
            (l.first_name = r.first_name and and l.last_name = r.last_name and l.dob = r.dob)

            where l."source_dataset" || '-__-' || l."unique_id" < r."source_dataset" || '-__-' || r."unique_id" and l."source_dataset" != r."source_dataset"
            )
        select
        count(*) as row_count,
        match_key
        from __splink__df_blocked_data
        group by match_key
        order by cast(match_key as int) asc
    )


Error was: Parser Error: syntax error at or near "and"
LINE 13: ...   CREATE TABLE __splink__df_count_cumulative_blocks_786f2c00a
        AS
    ...

**Reply by RobinL** [reply]
I think the model is iterating to a potentially strange result.

For example, in your settings you have:
```
 {
                    "sql_condition": "first_name_l = first_name_r",
                    "label_for_charts": "Exact match",
                    "m_probability": 1.0,
                    "u_probability": 0.001286008230452675,
                    "tf_adjustment_column": "first_name",
                    "tf_adjustment_weight": 1.0
                },
                {
                    "sql_condition": "levenshtein(first_name_l, first_name_r) <= 2",
                    "label_for_charts": "Levenstein <= 2",
                    "u_probability": 0.0033436213991769547,
                    "tf_adjustment_column": "first_name",
                    "tf_adjustment_weight": 0.5
                },
                {
                    "sql_condition": "ELSE",
                    "label_for_charts": "All other comparisons",
                    "m_probability": 3.287660633911516e-132,
                    "u_probability": 0.9953703703703703
                }
            ]
```

Which is saying that, amongst truly matching records, the first_name _never_ mismatches, and hence you have `  "m_probability": 3.287660633911516e-132,` in the ELSE level (an extremely tiny number) which might be the source of your problem.

Some thing to try:
- Using probability_two_random_records_match to set a more accurate parameter as mentioned above
- Adding SSN and MI to your list of comparisons.  More columns generally improves predictions


Nasty hacks which might work could be setting `"max_iterations" : 10` rather than 100, or manually overriding the 3.3e-132 m probability to something tiny but a bit bigger e.g. 1.0e-20

I realise this doesn't directly answer the question of why it was working previously - I'm not sure of the answer to that, but I don't think it's directly a result of the python version.

Am I right in saying you're linking fairly small datasets?  That can sometimes be the cause of 'extreme' parameter estimates.  

You could try loosening your training blocking rules to e.g. 
```
training_blocking_rule = "l.first_name = r.first_name"
training_session_fname_sname = linker.estimate_parameters_using_expectation_maximisation(training_blocking_rule)

training_blocking_rule2 = "l.surname= r.surname"
training_session_dob = linker.estimate_parameters_using_expectation_maximisation(training_blocking_rule2)
```

**Reply by RobinL** [reply]
That new error is coming from 
` (l.first_name = r.first_name and and l.last_name = r.last_name and l.dob = r.dob)`

I think you might have a duplicate 'and' in a blocking rule you're using  (it says `and and`)

**Reply by DOH-REP1303** [reply]
Thank you for your suggestions. I will try them out and then report back.

**Reply by DOH-REP1303** [reply]
Getting rid of the extra and helped. The code runs again. I will go ahead and make some of the other revisions you suggested. My test data is a small sample. The real data will have many more rows. Thank you so much. On to the next step of running this from SSIS and setting up branching for exact, inexact and no match.

**Reply by RobinL** [reply]
Great.  Yeah, that makes sense, it's common to get odd/extreme param estimates on small samples of data.  Usually models are more stable the more data they have to learn from

**Comment by DOH-REP1303** [comment]
Good Morning Robin,

I just wanted to take a moment to thank you again for all your help. It has really made a difference in helping me understand how Splink functions. I am now refining my specifications according to your guidance.

I hope you have a very pleasant weekend,

Rosalind

Rosalind Philips (she/her)
Public Health Systems Developer
Center for Technology Operations
Executive Office of Innovation and Technology
Washington State Department of Health
***@***.******@***.***>
360-236-4478| www.doh.wa.gov<http://www.doh.wa.gov/>
***@***.***<https://www.doh.wa.gov/Newsroom/SocialMedia>

***@***.***

**Reply by RobinL** [reply]
You're very welcome!  Always keen to support users working in important areas like health, do feel free to ask any other questions if they arise

**Comment by anilkgudla** [comment]
@RobinL I am getting the same error as well "Error was: Out of Range Error: cannot take logarithm of zero". Attaching the saved model, please let me know if you need more details on the error.

[saved_model_contacts.json.zip](https://github.com/moj-analytical-services/splink/files/12605845/saved_model_contacts.json.zip)
